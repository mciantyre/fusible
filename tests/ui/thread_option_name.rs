// SPDX-License-Identifier: MPL-2.0
// SPDX-FileCopyrightText: Copyright 2024 Ian McIntyre

use fusible::thread::{StaticStack, Thread, ThreadOptions};
use std::ffi::CString;
use std::pin::pin;

static STACK: StaticStack<512> = StaticStack::new();

/// A thread's context cannot outlive the name it captures.
fn main() {
    let thread = pin!(Thread::context());

    let name = CString::new("name").unwrap();
    let mut opts = ThreadOptions::default();
    opts.name = Some(&name);

    Thread::create(thread.into_ref(), STACK.take().unwrap(), &opts, || ()).unwrap();
}
