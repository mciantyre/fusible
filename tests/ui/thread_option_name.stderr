error[E0597]: `name` does not live long enough
  --> tests/ui/thread_option_name.rs:16:22
   |
14 |     let name = CString::new("name").unwrap();
   |         ---- binding `name` declared here
15 |     let mut opts = ThreadOptions::default();
16 |     opts.name = Some(&name);
   |                      ^^^^^ borrowed value does not live long enough
...
19 | }
   | -
   | |
   | `name` dropped here while still borrowed
   | borrow might be used here, when `pinned` is dropped and runs the `Drop` code for type `ThreadContext`
   |
   = note: values in a scope are dropped in the opposite order they are defined
